### PR TITLE
UI for defining parameters

### PR DESCRIPTION
**Pull request by ahgittin**
has a field-based mode and JSON mode

builds on #118 , makes it mergeable

ideally flaws in the PR can be tracked here but addressed in follow-on PRs unless critical.  i'm aware of the following wishlist items:

1. if no params, it's not obvious how to add ("No params" should have option to "Add" one)
2. are the icons in top-right of params section needed (filter etc)
3. if same parameter entered twice you get angular errors
4. parameters section should appear only on application (it's pointless elsewhere)
5. parameters added should show up in the config section (maybe?)
6. do all the fields work as expected when defining a parameter used subsequently (eg label, constraint)
7. parameters are mostly pointless if you're deploying directly, it's mainly for saving and reuse (or some types of export)

i think 1,4,6 need addressed imminently but i don't think they should block merging this

### CODE REVIEW

**Comment by ahgittin** [comment]
i've addressed 1,2,4, and applied a similar toolbar cleanup to configuration (also use labels a la palette instead of checkboxes for the filters).  3 is minimised due to better duplicate prevention logic.

needs some testing for 5 and 6 (but shouldn't block this merge)

image

**Comment by ahgittin** [comment]
did some testing and it all works (6 is happy), except a default value of `''` is being set for things so `required` has no effect.

i've fixed that such that we drop fields if empty (including constraints) except for default value we're slightly more clever if `""` is set in yaml; see last commit for details.

this leaves two things to consider:

* 5 -- i can't think of a use case for including params in the config section, so while logically it is valid to include them, practically i see no reason to, it just makes noise and confusion; setting a `default` is clearer than setting a value and they achieve the same things
* 7 -- not one for now but as mentioned in #118 we might want to merge some of the catalog metadata into the "Application" widget instead of the modal "Save" popup, with an option there "Save this as a catalog item" and parameters only appears if that is checked

**Comment by jcabrerizo** [comment]
Great job you both! I'd like add some ideas:

1. When we add an efector to an application, it shows an small icon on the top right. Have something similar to notice the user that this app has properties woudl help him
image
2. Parameter type is a text input box. It fails when we tray to add something out of [string, bool, boolean, byte, char, character, short, integer, int, long, float, double, duration, timestamp, port], a select input seem to be the correct here
3. Same to the above to Constraints
4. Even names are self-explanatory, I'd add an icon with info to, types, and default value to inform it accepts an empty string

None of this is a blocker so we can do it in future

**Comment by ahgittin** [comment]
all good suggestions @jcabrerizo , TY.  your 2 and 4 are easy, i'll do that in this PR.

not sure about 1, it might make the display noisy.  but we should consider as part of a wider UI review at some point.

i'd like to do a lot more around constraints but there aren't quick wins, as it needs a list and some of the constraints themselves take parameters (eg `requiredUnless("otherKey")`).

**Comment by ahgittin** [comment]
@jcabrerizo both updated; went with a subset of common parameter types, with note in code and ability to write others

good to merge in your opinion?

**Comment by jcabrerizo** [comment]
Yes, looks good to me now except for one weird thing: you can not use `name` as key name. Makes sense?

**Comment by ahgittin** [comment]
that's weird, `name` seems to make it blow up.  i'll look at that.

**Comment by ahgittin** [comment]
took me far too long to find, but i've got it now @jcabrerizo .  can you look at last commit and then we can merge!

**Comment by jcabrerizo** [comment]
Perfect now 👍

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 1]
```diff
@@ -449,13 +527,25 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
...
+            return item.widgetMode;
+        };
+        specEditor.getParameterWidgetMode = (item) => {
+            let type = item.type || item.typeName;
+
+            if (type === 'java.lang.Boolean') type = 'boolean';
+            else if (type === 'java.util.Map') type = 'map';
+            else if (type === 'java.util.Set' || type === 'java.util.List' || type === 'java.util.Collection' || type.startsWith('List<')) type = 'array';
+
+            if (type.startsWith('AWS::')) type = 'unknown';
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]: Why is there a mention of AWS here?

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 2]
```diff
@@ -503,7 +593,7 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
             if (!specEditor.defined(value)) {
                 value = null;
             }
-            if (value!=null) {
+            if (value != null) {
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]:
```suggestion
            if (value !== null) {
```

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 3]
```diff
@@ -548,24 +638,65 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
                         value = JSON.stringify(value);
                     }
                 }
-                if (value!=null) {
+                if (value != null) {
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]: Should use `!==`
```suggestion
                if (value !== null) {
```

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 4]
```diff
@@ -548,24 +638,65 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
...
                     scope.config[item.name] = value;
                 }
             }
             scope.state.config.codeModeActive[item.name] = !oldMode;
-            if (value!=null) {
+            if (value != null) {
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]:
```suggestion
            if (value !== null) {
```

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 5]
```diff
@@ -548,24 +638,65 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
...
             } else {
-                return "Edit in simple mode, unwrapping JSON if possible ["+itemName+"]";
+                // entering code mode
+                // leave code mode for other parameter if set
+                if (scope.state.parameters.edit.newName && scope.state.parameters.edit.newName!=item.name) {
+                    scope.parameterCodeModeClick(getParameter(scope.state.parameters.edit.newName));
+                }
+                // convert local parameter from json to non-json or vice-versa
+                value = item;
+                if (value != null) {
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]:
```suggestion
                if (value !== null) {
```

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 6]
```diff
@@ -601,36 +738,37 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
...
+                    $templateCache.put(templateName, '<i>Widget ' + widgetName + ' missing</i>');
                 }
             }
             return templateName;
         };
-
         specEditor.isDsl = (key, index) => {
             let val = scope.model.config.get(key);
-            if (specEditor.defined(val) && specEditor.defined(index) && index!=null) val = val[index];
+            if (specEditor.defined(val) && specEditor.defined(index) && index != null) val = val[index];
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]:
```suggestion
            if (specEditor.defined(val) && specEditor.defined(index) && index !== null) val = val[index];
```

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 7]
```diff
@@ -639,9 +777,9 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
         };
         specEditor.isDslWizardButtonAllowed = (key, index, nonModelValue) => {
             let val = scope.model.config.get(key);
-            if (specEditor.defined(val) && specEditor.defined(index) && index!=null) val = val[index];
-            if (!specEditor.defined(val) || val===null || val==='') val = nonModelValue;
-            if (!specEditor.defined(val) || val===null || val==='') return true;
+            if (specEditor.defined(val) && specEditor.defined(index) && index != null) val = val[index];
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]:
```suggestion
            if (specEditor.defined(val) && specEditor.defined(index) && index !== null) val = val[index];
```

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 8]
```diff
@@ -654,48 +792,61 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
...
                 }
-
                 result[key] = getLocalConfigValueFromModelValue(key, value);
             }
             scope.config = result;
         }
 
         function getLocalConfigValueFromModelValue(key, value) {
-            if (!specEditor.defined(value) || value==null) {
+            if (!specEditor.defined(value) || value == null) {
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]:
```suggestion
            if (!specEditor.defined(value) || value === null) {
```

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 9]
```diff
@@ -788,11 +938,11 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
             }
 
             // all other primitives treat as string (as they will get a string-based widget)
-            return ""+value;
+            return "" + value;
         }
 
         function getModelValueFromString(val) {
-            if (!specEditor.defined(val) || val==null || typeof val !== 'string') {
+            if (!specEditor.defined(val) || val == null || typeof val !== 'string') {
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]:
```suggestion
            if (!specEditor.defined(val) || val === null || typeof val !== 'string') {
```

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 10]
```diff
@@ -888,13 +1037,151 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
...
+                        } else {
+                            scope.state.parameters.edit.errors.push({ message: "Constraint JSON must be a list" });
+                        }
+                    } catch (e) {
+                        // $log.warn("ERROR parsing constraints", scope.state.parameters.edit.constraints, e);
+                        scope.state.parameters.edit.errors.push({ message: "Invalid constraint JSON" });
+                    }
+                    
+                    // empty values are removed
+                    if (item.description == '') {
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]:
```suggestion
                    if (item.description === '') {
```

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 11]
```diff
@@ -888,13 +1037,151 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
...
+                        delete item['description'];
+                    } 
+                    if (item.label == '') {
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]:
```suggestion
                    if (item.label === '') {
```

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js [hunk 12]
```diff
@@ -888,13 +1037,151 @@ export function specEditorDirective($rootScope, $templateCache, $injector, $sani
...
+                        delete item['label'];
+                    } 
+                    if (item.default == '') {
```
Comment (tbouron) [inline ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]:
```suggestion
                    if (item.default === '') {
```

File: ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js (hunk 1 above)
Comment (ahgittin) [reply ui-modules/blueprint-composer/app/components/spec-editor/spec-editor.directive.js]: quite right, this shouldn't be here!

**Comment by ahgittin** [comment]
great points @tbouron 

> `==` vs `===` and `!=` vs `!==` ...

you are 100% correct we should use the latter.  i revert to bad other language habits.  TY.

> (3) is absolutely critical if you have duplicate key. We can live with awkward behaviour but not breaking the UI. This can be solved by adding validation on the key field, to check if any exists already.

yeah, i buy that.  i don't see an easy way to do it with validation for the case where duplicate keys are set in yaml.  (we already did it if you tried to supply duplicate keys in the ui so i think from yaml is the only way.)  what i've done instead is that on model load if we find a duplicate we append a number to it, so `brooklyn.parameters: [ { name: y, type: string }, { name: y, type: string }, { name: y, type: string } ]` becomes `y` `y2` `y3`, with warnings logged to the console.  it's awkward and surprising but at least doesn't break the ui.  we could have yaml model validation errors as well for this though the problem would still manifest in the graphical composer if invalid yaml was passed in so we need a way to handle it here.  the other option would be to refuse to show parameters.

> any constraints added into the the constraint field don't work.  Worse, the value is cleared as soon as I unfocus the field which is very irritating if the constraint is long. And setting the constraints in either JSON mode or directly in YAML will silently remove them.

if it's valid JSON it works, but it must be e.g. `["required"]` -- -have changed message to note it needs the brackets _and_ double quotes.  could improve by supporting yaml parsing or a dedicated editor but going for minimal working capability here.  also agree clearing on unfocus is awkward but of course there's a lot of work to store invalid models if we did something else.  we could maybe forbid the unfocus.

> the change to have a single filter/add field is moot. It's awkward to use and hard to understand what it's doing. This would be high priority to fix IMO but in the sake of getting this PR merged (and the spirit of my first point) it can be addressed in a subsequent PR.

doesn't moot mean "no longer relevant or not worth discussion"?  i think you mean it's a step backwards.  personally i thought having two was awkward and one is fairly self-explanatory.  happy to see further evolution and testing but agree it's such a minor point as to be _moot_ :).

> re the new control for filter/add config and parameter, I think the icons you added are unnecessary in this case: it takes valuable real estate and don't give meaning. For example, the cross for required will be interpreted as "close" or "remove" the required params. I think i would be best to not include them for now, until we find a better way.

i removed checkboxes so in total we have a real estate savings.  i think icons make it look more polished, but agree the `x` is confusing.  have changed it to a `stop-circle` (bit white square in black circle).  again happy to see further evolution but i'd like to keep the icons.  if you feel strongly let's get a tie-break opinion!

**Comment by ahgittin** [comment]
(all comments addressed and pushed @tbouron )

**Comment by tbouron** [comment]
>yeah, i buy that. i don't see an easy way to do it with validation for the case where duplicate keys are set in yaml. (we already did it if you tried to supply duplicate keys in the ui so i think from yaml is the only way.) what i've done instead is that on model load if we find a duplicate we append a number to it, so brooklyn.parameters: [ { name: y, type: string }, { name: y, type: string }, { name: y, type: string } ] becomes y y2 y3, with warnings logged to the console. it's awkward and surprising but at least doesn't break the ui. we could have yaml model validation errors as well for this though the problem would still manifest in the graphical composer if invalid yaml was passed in so we need a way to handle it here. the other option would be to refuse to show parameters.

Sounds good for now 👍 

> if it's valid JSON it works, but it must be e.g. ["required"] -- -have changed message to note it needs the brackets and double quotes. could improve by supporting yaml parsing or a dedicated editor but going for minimal working capability here. also agree clearing on unfocus is awkward but of course there's a lot of work to store invalid models if we did something else. we could maybe forbid the unfocus.

Aaaah it requires JSON! Great for the error message, **it would be also great to specify this in the information popup for the constraints @ahgittin.**

> doesn't moot mean "no longer relevant or not worth discussion"? i think you mean it's a step backwards. personally i thought having two was awkward and one is fairly self-explanatory. happy to see further evolution and testing but agree it's such a minor point as to be moot :).

Probably my flawed French translation: `moot` means `subject to debate, arguable` to me. 

> i removed checkboxes so in total we have a real estate savings. i think icons make it look more polished, but agree the x is confusing. have changed it to a stop-circle (bit white square in black circle). again happy to see further evolution but i'd like to keep the icons. if you feel strongly let's get a tie-break opinion!

Great!

Happy to merge once we have the explanation for the JSON constraints @ahgittin !

**Comment by ahgittin** [comment]
@tbouron the constraints hover text currently says:

> Constraints such as "required" or "forbiddenUnless(otherKey)", expressed as a JSON list (wrapped in square brackets and double quotes)

is this enough?  or did you want more explanation did you have in mind?

would be if we made it html and could use code formatting, we could say

> Constraints such as `[ "required" ]` or `[ "forbiddenUnless(otherKeyNeededIfAndOnlyIfThisOne)", "requiredUnless(otherKeyNeededIfAndOnlyIfThisOne)" ]`, expressed as a JSON list (wrapped in square brackets and double quotes)

but it's fiddly to make it consistent with the others while supporting html so i'm inclined to leave it as is

**Comment by tbouron** [comment]
> > Constraints such as "required" or "forbiddenUnless(otherKey)", expressed as a JSON list (wrapped in square brackets and double quotes)

> is this enough? or did you want more explanation did you have in mind?

Yes, should be enough, happy to merge

**Pull request merged**